Mathlib/CategoryTheory/MorphismProperty/Lifting.lean

```diff
@@ -0,0 +1,159 @@
+  ∀ ⦃X Y : C⦄ ⦃g : X ⟶ Y⦄ (_ : T g), HasLiftingProperty g f
+
+lemma Llp.IsStableUnderRetracts {T : MorphismProperty C} : T.Llp.IsStableUnderRetracts := by
+  refine ⟨fun {X Y} _ _ f f' H L _ _ g h ↦ ⟨fun {u v} sq ↦ ?_⟩⟩
+  have : (H.r.left ≫ u) ≫ g = f' ≫ (H.r.right ≫ v) := by simp [sq.w]
+  obtain lift := ((L h).sq_hasLift (CommSq.mk this)).exists_lift.some
+  refine ⟨H.i.right ≫ lift.l, ?_, ?_⟩
+  · have h : H.i.left ≫ f' = f ≫ H.i.right := H.i.w
+    have h' : H.i.left ≫ H.r.left = 𝟙 X := Arrow.hom.congr_left H.retract
+    rw [← Category.assoc, ← h, Category.assoc, lift.fac_left, ← Category.assoc, h',
+      Category.id_comp]
+  · have : H.i.right ≫ H.r.right = 𝟙 Y := Arrow.hom.congr_right H.retract
+    rw [Category.assoc, lift.fac_right, ← Category.assoc, this, Category.id_comp]
+
+lemma Rlp.IsStableUnderRetracts {T : MorphismProperty C} : T.Rlp.IsStableUnderRetracts := by
+  refine ⟨fun {X Y} _ _ f f' H L _ _ g h ↦ ⟨fun {u v} sq ↦ ?_⟩⟩
+  have : (u ≫ H.i.left) ≫ f' = g ≫ (v ≫ H.i.right) := by
+    rw [← Category.assoc, ← sq.w]
+    aesop
+  obtain lift := ((L h).sq_hasLift (CommSq.mk this)).exists_lift.some
+  refine ⟨lift.l ≫ H.r.left, ?_, ?_⟩
+  · have h : H.i.left ≫ H.r.left = 𝟙 X := Arrow.hom.congr_left H.retract
+    rw [← Category.assoc, lift.fac_left, Category.assoc, h, Category.comp_id]
+  · have h : H.r.left ≫ f = f' ≫ H.r.right := H.r.w
+    have h' : H.i.right ≫ H.r.right = 𝟙 Y := Arrow.hom.congr_right H.retract
+    rw [Category.assoc, h, ← Category.assoc, lift.fac_right, Category.assoc, h', Category.comp_id]
+
+open Limits.PushoutCocone in
+instance Llp.IsStableUnderCobaseChange {T : MorphismProperty C} :
+    T.Llp.IsStableUnderCobaseChange := by
+  refine ⟨fun {A B A' B'} f s f' t P L X Y g hg ↦ ⟨fun {u v} sq ↦ ?_⟩⟩
+  have w : (s ≫ u) ≫ g = f ≫ (t ≫ v) := by
+    rw [← Category.assoc, ← P.toCommSq.w, Category.assoc, Category.assoc, sq.w]
+  obtain lift := ((L hg).sq_hasLift (CommSq.mk w)).exists_lift.some
+  let lift' : B' ⟶ X := IsColimit.desc P.isColimit u lift.l (by rw [lift.fac_left])
+  let l : f' ≫ lift' = u := IsColimit.inl_desc P.isColimit u lift.l (by rw [lift.fac_left])
+  let l' : t ≫ lift' = lift.l := IsColimit.inr_desc P.isColimit u lift.l (by rw [lift.fac_left])
+  let newCocone := mk (f' ≫ v) (t ≫ v) (by have := P.w; apply_fun (fun f ↦ f ≫ v) at this; aesop)
+  refine ⟨lift', l,
+    (P.isColimit.uniq newCocone (lift' ≫ g) ?_).trans (P.isColimit.uniq newCocone v ?_).symm⟩
+  all_goals dsimp [newCocone]; intro j; cases j; simp only [Limits.span_zero, condition_zero,
+    IsPushout.cocone_inl, Category.assoc]
+  · rw [← Category.assoc, ← Category.assoc, Category.assoc s f' lift', l, ← sq.w, Category.assoc]
+  · rename_i k; cases k; all_goals dsimp
+    · rw [← Category.assoc, l, sq.w]
+    · rw [← Category.assoc, l', lift.fac_right]
+  · rename_i k; cases k; all_goals dsimp
+
+open Limits.PullbackCone in
+instance Rlp.IsStableUnderBaseChange {T : MorphismProperty C} :
```

Comment:
You may also add a `IsStableUnderProductsOfShape` lemma (and also the dual `IsStableUnderCoproductsOfShape`, even though the latter is not defined yet in `MorphismProperty.Limits`).